jarbas/dashboard/admin/__init__.py

```diff
@@ -149,4 +155,80 @@ def get_search_results(self, request, queryset, search_term):
         return queryset, distinct
 
 
+class ReimbursementSummaryModelAdmin(PublicAdminModelAdmin):
+    change_list_template = 'dashboard/reimbursement_summary_change_list.html'
+    date_hierarchy = 'issue_date'
+
+    def get_next_in_date_hierarchy(self, request):
+        if f'{self.date_hierarchy}__month' in request.GET:
+            return 'day'
+        if f'{self.date_hierarchy}__year' in request.GET:
+            return 'month'
+        return 'year'
+
+    @staticmethod
+    def bar_chart_height(row, low, high):
+        total, low, high = row['total'] or 0, low or 0, high or 0
+        if low == high:
+            size = 0
+        else:
+            size = (total - low) / (high - low) * 100
+
+        corrected_size = Decimal('5') + (Decimal('0.95') * size)
+        return corrected_size
+
+    def changelist_view(self, request, extra=None):
+        response = super().changelist_view(request, extra_context=extra)
+
+        try:
+            queryset = response.context_data['cl'].queryset
```

Comment:
Since I wrote the article I found that Django will execute two more unnecessary queries on the underlying model (in your case it's `Reimbursement`). If the model holds a lot of data, these queries can be eliminated, but it requires some dirty hacks. 

If you find performance to be an issue in this specific page, you can eliminate the two extra queries.

The two queries original from [within the Django admin](https://github.com/django/django/blob/master/django/contrib/admin/options.py#L1780). To eliminate them you need to fool Django into thinking the queryset is empty. But, you still want Django to pass the queryset through the `ChangeList` and apply all the filters (from `date_hierarchy` and `list_filters`).

First, you need to override the `get_queryset` function of the `ReimbursementSummaryModelAdmin` class:

```python
def get_queryset(self, request):
        return super().get_queryset(request)
```

To fool Django into thinking the queryset is empty,  you need to return a queryset that will return no results, but one that you can reverse later (after the filters were applied):

```python
def add_false_condition(qs):
    return qs.extra(where=['1=0'])

...

class ReimbursementSummaryModelAdmin(PublicAdminModelAdmin): 
   ...
   def get_queryset(self, request):
        return add_false_condition(super().get_queryset(request).using(settings.READ_REPLICA_USING))
```

This will add the false condition "1=0" to the query. Some vendors, such as PostgreSQL will not do anything when they see such condition. This is good, this is what we want.

Now, you still need the queryset with all it's filters when you fetch the data for the list view. This is where we remove the fals condition:

```python
def remove_false_condition(qs):
    clone = qs.all()
    clone.query.where.children = [
        e for e in qs.query.where.children
        if not hasattr(e, 'sqls') or e.sqls != ['1=0']
    ]
    return clone

...

class ReimbursementSummaryModelAdmin(PublicAdminModelAdmin): 
   ...
   def changelist_view(self, request, extra_context=None):
       response = super().changelist_view(request, extra_context=extra_context)

      try:
          qs = response.context_data['cl'].queryset
      except (AttributeError, KeyError):
          return response

       qs = remove_false_condition(qs)
```

Feel free to use this but be aware that this is a really dirty hack. I only had to use this hack after our app got *really* big.

Reply:
We do have a bottleneck in Postgres, so we use a some caching strategies. For this summary I'm caching the whole context we calculate in b0c4ab3. That should be enough, but gonna monitor performance once it rolls out.